File: go/pii_in_detail_test.go
package fasten

import (
	"context"
	"database/sql"
	"strings"
	"testing"

	// Pure-Go SQLite driver — no CGO required so this test runs under
	// CGO_ENABLED=0 CI (e.g. GitHub Actions default).
	_ "modernc.org/sqlite"
)

// resetForPII clears state set by other tests in this package.
func resetForPII(t *testing.T) {
	t.Helper()
	regMu.Lock()
	for k := range _registry {
		if strings.HasPrefix(string(k), "PII_") || strings.HasPrefix(string(k), "NON_PII_") {
			delete(_registry, k)
		}
	}
	regMu.Unlock()
}

// ── #1 force RetShort at registration ────────────────────────────────────

func TestPII_ForcesRetShort(t *testing.T) {
	resetForPII(t)
	if err := Register("pii", map[Code]Meta{
		"PII_FORCE_SHORT": {
			Domain:         "pii", Category: "x", Action: "x",
			Severity:       SevInfo,
			Description:    "x", Emitter: "t",
			RetentionClass: RetLong,
			PiiInDetail:    true,
		},
	}); err != nil {
		t.Fatalf("Register: %v", err)
	}
	m, _ := metaOf("PII_FORCE_SHORT")
	if m.RetentionClass != RetShort {
		t.Fatalf("expected RetShort, got %v", m.RetentionClass)
	}
}

// ── #2 force-redact detail on emit ───────────────────────────────────────

func TestPII_ForceRedactsDetail(t *testing.T) {
	resetForPII(t)
	resetGlobals(t)
	_ = Init(Config{ServiceID: "svc", NodeID: "node"})
	if err := Register("pii", map[Code]Meta{
		"PII_FULL_REDACT": {
			Domain:    "pii", Category: "c", Action: "a",
			Severity:  SevInfo,
			Description: "x", Emitter: "t",
			PiiInDetail: true,
		},
	}); err != nil {
		t.Fatalf("Register: %v", err)
	}

	row, err := Emit(context.Background(), "PII_FULL_REDACT",
		Target("u-42"),
		WithDetail(map[string]any{"city": "Mumbai", "ip": "203.0.113.1"}),
	)
	if err != nil {
		t.Fatalf("Emit: %v", err)
	}
	if _, ok := row.Detail["city"]; ok {
		t.Fatal("city must not survive PII force-redact")
	}
	if _, ok := row.Detail["ip"]; ok {
		t.Fatal("ip must not survive PII force-redact")
	}
	if row.Detail["_redacted"] != "***" {
		t.Fatalf("expected _redacted='***', got %v", row.Detail["_redacted"])
	}
	if row.Detail["_pii_in_detail"] != true {
		t.Fatal("_pii_in_detail must be true")
	}
	if !row.PiiInDetail {
		t.Fatal("Row.PiiInDetail must be true for PII codes")
	}
}

func TestPII_PassthroughKeys(t *testing.T) {
	resetForPII(t)
	resetGlobals(t)
	_ = Init(Config{ServiceID: "svc", NodeID: "node"})
	if err := Register("pii", map[Code]Meta{
		"PII_PARTIAL": {
			Domain:    "pii", Category: "c", Action: "a",
			Severity:  SevInfo,
			Description: "x", Emitter: "t",
			PiiInDetail:           true,
			DetailPassthroughKeys: []string{"region"},
		},
	}); err != nil {
		t.Fatalf("Register: %v", err)
	}
	row, err := Emit(context.Background(), "PII_PARTIAL",
		Target("u-42"),
		WithDetail(map[string]any{
			"city":    "Mumbai",
			"region":  "MH",
			"api_key": "sk-secret",
		}),
	)
	if err != nil {
		t.Fatalf("Emit: %v", err)
	}
	if row.Detail["region"] != "MH" {
		t.Fatalf("expected region=MH, got %v", row.Detail["region"])
	}
	if _, ok := row.Detail["city"]; ok {
		t.Fatal("city must not survive (not in passthrough)")
	}
	// api_key not in passthrough list, so it's gone entirely
	if _, ok := row.Detail["api_key"]; ok {
		t.Fatal("api_key must not survive (not in passthrough)")
	}
}

func TestNonPII_KeepsDetail(t *testing.T) {
	resetForPII(t)
	resetGlobals(t)
	_ = Init(Config{ServiceID: "svc", NodeID: "node"})
	if err := Register("nonpii", map[Code]Meta{
		"NON_PII_BASIC": {
			Domain:    "nonpii", Category: "c", Action: "a",
			Severity:  SevInfo,
			Description: "x", Emitter: "t",
		},
	}); err != nil {
		t.Fatalf("Register: %v", err)
	}
	row, err := Emit(context.Background(), "NON_PII_BASIC",
		Target("u-42"),
		WithDetail(map[string]any{"city": "Mumbai", "api_key": "sk-secret"}),
	)
	if err != nil {
		t.Fatalf("Emit: %v", err)
	}
	if row.Detail["city"] != "Mumbai" {
		t.Fatal("non-secret key must survive non-PII codes")
	}
	if row.Detail["api_key"] != "***" {
		t.Fatal("secret key must still be redacted")
	}
	if _, ok := row.Detail["_pii_in_detail"]; ok {
		t.Fatal("non-PII row must not carry _pii_in_detail marker")
	}
	if row.PiiInDetail {
		t.Fatal("Row.PiiInDetail must be false for non-PII codes")
	}
}

// ── #3 SQLite column persists pii_in_detail ──────────────────────────────

func TestPII_SQLitePersistsColumn(t *testing.T) {
	db, err := sql.Open("sqlite", ":memory:")
	if err != nil {
		t.Fatalf("sqlite open: %v", err)
	}
	defer db.Close()

	store, err := NewSQLiteStore(db, "audit_pii_test")
	if err != nil {
		t.Fatalf("NewSQLiteStore: %v", err)
	}

	row := Row{
		ID: "evt-pii", OriginID: "evt-pii", MonotonicSeq: 1,
		Code: "X_PII", Action: "x", Severity: SevInfo,
		ServiceID: "s", SourceNodeID: "n",
		Actor: "a", ActorKind: "user",
		Target: "t", Category: "c", Domain: "d",
		Method: "sdk", RequestID: "abc123def456",
		Detail:      map[string]any{"_redacted": "***"},
		PiiInDetail: true,
	}
	if err := store.Insert(context.Background(), row); err != nil {
		t.Fatalf("Insert: %v", err)
	}
	got, err := store.Query(context.Background(), Filter{RequestID: "abc123def456"})
	if err != nil {
		t.Fatalf("Query: %v", err)
	}
	if len(got) != 1 || !got[0].PiiInDetail {
		t.Fatalf("expected one row with PiiInDetail=true, got %+v", got)
	}
}

File: js/test/pii_in_detail.test.mjs
/**
 * P1-5: pii_in_detail enforcement (JS parity with Python + Go).
 *
 * 1. register() forces retentionClass=SHORT and warns when piiInDetail=true
 * 2. emit() force-redacts detail; passthrough keys survive (still scrubbed)
 * 3. row.pii_in_detail mirrors meta.piiInDetail
 */
import { test, beforeEach, describe } from 'node:test';
import assert from 'node:assert/strict';
import {
    init, emit, register, metaOf,
    REDACT_REPLACEMENT, RetentionClass,
    registry, _clearRustRegistry,
} from '../src/index.js';

function clear() {
    _clearRustRegistry();
    registry().clear();
}

// Capture console.warn calls.
function captureWarn(fn) {
    const original = console.warn;
    const calls = [];
    console.warn = (...args) => calls.push(args.join(' '));
    try { fn(); } finally { console.warn = original; }
    return calls;
}

// ── #1 force SHORT at registration ────────────────────────────────────────

describe('P1-5 register: piiInDetail forces SHORT', () => {
    beforeEach(clear);

    test('forces SHORT and warns when adopter set LONG', () => {
        const warnings = captureWarn(() => {
            register('pii', {
                PII_LONG_OVERRIDE: {
                    domain: 'pii', category: 'p', action: 'view',
                    severity: 'info', description: 'x', emitter: 't',
                    piiInDetail: true,
                    retentionClass: RetentionClass.LONG,
                },
            });
        });
        assert.equal(metaOf('PII_LONG_OVERRIDE').retentionClass, RetentionClass.SHORT);
        assert.ok(
            warnings.some(w => w.includes('PII_LONG_OVERRIDE') && w.includes('forced to SHORT')),
            `expected warning about forced override; got: ${warnings.join(' | ')}`,
        );
    });

    test('no warning when adopter already declared SHORT', () => {
        const warnings = captureWarn(() => {
            register('pii', {
                PII_ALREADY_SHORT: {
                    domain: 'pii', category: 'p', action: 'view',
                    severity: 'info', description: 'x', emitter: 't',
                    piiInDetail: true,
                    retentionClass: RetentionClass.SHORT,
                },
            });
        });
        assert.equal(
            warnings.filter(w => w.includes('PII_ALREADY_SHORT') && w.includes('forced')).length,
            0,
        );
    });

    test('non-PII code keeps adopter retentionClass', () => {
        register('pii', {
            NON_PII_LONG: {
                domain: 'pii', category: 'p', action: 'view',
                severity: 'info', description: 'x', emitter: 't',
                piiInDetail: false,
                retentionClass: RetentionClass.LONG,
            },
        });
        assert.equal(metaOf('NON_PII_LONG').retentionClass, RetentionClass.LONG);
    });
});

// ── #2 force-redact + passthrough at emit ─────────────────────────────────

describe('P1-5 emit: force-redact + passthrough', () => {
    beforeEach(() => {
        clear();
        init({ serviceId: 'svc', nodeId: 'node' });
    });

    test('whole detail is force-redacted for PII codes', () => {
        register('pii', {
            PII_FULL_REDACT: {
                domain: 'pii', category: 'p', action: 'view',
                severity: 'info', description: 'x', emitter: 't',
                piiInDetail: true,
            },
        });
        const row = emit({
            code: 'PII_FULL_REDACT', target: 'u-42',
            detail: { city: 'Mumbai', ip: '203.0.113.1' },
        });
        assert.ok(!('city' in row.detail), 'city must not survive force-redact');
        assert.ok(!('ip' in row.detail), 'ip must not survive force-redact');
        assert.equal(row.detail._redacted, REDACT_REPLACEMENT);
        assert.equal(row.detail._pii_in_detail, true);
        assert.equal(row.pii_in_detail, true);
    });

    test('passthrough keys survive', () => {
        register('pii', {
            PII_PARTIAL: {
                domain: 'pii', category: 'p', action: 'view',
                severity: 'info', description: 'x', emitter: 't',
                piiInDetail: true,
                detailPassthroughKeys: ['region'],
            },
        });
        const row = emit({
            code: 'PII_PARTIAL', target: 'u-42',
            detail: { city: 'Mumbai', region: 'MH', api_key: 'sk-secret' },
        });
        assert.equal(row.detail.region, 'MH');
        assert.ok(!('city' in row.detail));
        assert.ok(!('api_key' in row.detail), 'api_key not in passthrough → gone entirely');
        assert.equal(row.detail._redacted, REDACT_REPLACEMENT);
    });

    test('passthrough still runs secret-key redactor on the value', () => {
        register('pii', {
            PII_PASSTHROUGH_SECRET: {
                domain: 'pii', category: 'p', action: 'view',
                severity: 'info', description: 'x', emitter: 't',
                piiInDetail: true,
                detailPassthroughKeys: ['api_key'],
            },
        });
        const row = emit({
            code: 'PII_PASSTHROUGH_SECRET', target: 'u-42',
            detail: { api_key: 'sk-secret' },
        });
        assert.equal(row.detail.api_key, REDACT_REPLACEMENT);
    });

    test('non-PII code keeps detail (only secret keys redacted)', () => {
        register('nonpii', {
            NON_PII_BASIC: {
                domain: 'nonpii', category: 'p', action: 'view',
                severity: 'info', description: 'x', emitter: 't',
            },
        });
        const row = emit({
            code: 'NON_PII_BASIC', target: 'u-42',
            detail: { city: 'Mumbai', api_key: 'sk-secret' },
        });
        assert.equal(row.detail.city, 'Mumbai');
        assert.equal(row.detail.api_key, REDACT_REPLACEMENT);
        assert.ok(!('_redacted' in row.detail));
        assert.ok(!('_pii_in_detail' in row.detail));
        assert.equal(row.pii_in_detail, false);
    });
});
